### DISCUSSION TITLE
OUTLOOK update MFA

### DISCUSSION THREAD

**Discussion by dividduang** [ecederstrand/exchangelib#1150] [Q&A] [answered]
Hi Erik,
Our company updated the new authentication method, which caused my program error
``
    raise AutoDiscoverFailed(
    exchangelib.errors.AutoDiscoverFailed: All steps in the autodiscover protocol failed for email xxx@xxxx.com. If you think
    this is an error, consider doing an official test at https://testconnectivity.microsoft.com

What should I do？
thx!!!!!

**Comment by dividduang** [comment]
https://learn.microsoft.com/zh-cn/azure/active-directory/authentication/how-to-migrate-mfa-server-to-azure-mfa?WT.mc_id=Portal-Microsoft_AAD_IAM

**Comment by ecederstrand** [comment]
You probably need to update your credentials setup to use OAuth or whatever your server accepts now. Additionally, here are some suggestions you could try if autodiscover still fails with OAuth: https://github.com/ecederstrand/exchangelib/issues/1146#issuecomment-1332404489

**Comment by dividduang** [comment]
Thank you very much for your reply, how can i  get it 'MY_ID''MY_SECRET','TENANT_ID'
credentials = OAuth2Credentials(client_id='MY_ID', client_secret='MY_SECRET', tenant_id='TENANT_ID')

**Reply by ecederstrand** [reply]
There's a guide for Office365 at https://ecederstrand.github.io/exchangelib/#oauth-on-office-365 but you may need to ask your Exchange admins to register an app for you.

**Reply by dividduang** [reply]
I've got these three ID ，But I'm still not going to change it code  ，how can i use exchangelib send email ?
```python
from exchangelib import DELEGATE, Account, Credentials, Message, Mailbox, HTMLBody,FileAttachment,Configuration,NTLM
from exchangelib.protocol import BaseProtocol,NoVerifyHTTPAdapter
from exchangelib import Configuration, OAuth2Credentials, OAuth2AuthorizationCodeCredentials, Identity, OAUTH2

def Email(to,cc_mail,subject, body, attachmentspath=None):
    creds = Credentials(
        username='xxxxxx@xxxxxxx.com',  
        password='***********'  
    )

    conf = Configuration(credentials=credentials, auth_type=OAUTH2)
    account = Account(
        primary_smtp_address='xxxxx@xxxxxx.com',  
        credentials=creds,
        autodiscover=False,
        access_type=DELEGATE,
        config = conf,

    )
    m = Message(
        account=account,
        cc_recipients=cc_mail,
        subject=subject,
        body=HTMLBody(body),
        to_recipients=[Mailbox(email_address=to)]
    )
```

I think that's the problem for most people，
THX!!!!!!!!!!!!THANK VERY MUCH

**Reply by ecederstrand** [reply]
Remove the `Credentials` object and don't pass it to `Account`. Your `Configuration` object already has the OAuth credentials you need for authentication.

**Reply by dividduang** [reply]
Yes, my email has been authorized. There is no problem with stmplib

**Comment by dividduang** [comment]
raise AutoDiscoverFailed(
    exchangelib.errors.AutoDiscoverFailed: All steps in the autodiscover protocol failed for email xxxxx@xxxxxxx.com. If you think 
    this is an error, consider doing an official test at https://testconnectivity.microsoft.com

Keep making error
 😭

**Reply by ecederstrand** [reply]
You set `autodiscover=False` in the above code, so exchangelib shouldn't be doing autodiscover at all. Please post the code that gave you the above `AutodiscoverFailed` error.

If you want to do autodiscovery with the OAuth type you are using, you need to set `autodiscover="soap"`, as described in https://github.com/ecederstrand/exchangelib/issues/1146#issuecomment-1332404489

**Reply by dividduang** [reply]
New Developments：It was ours IT   No Settings  ’https://github.com/robocorp/example-oauth-email#microsoft-exchange-outlook‘ Delegated: (Office 365 Exchange Online) EWS.AccessAsUser.All   <<<<But I didn't find a way to set it up ，how can i set       EWS.AccessAsUser.All???

**Reply by dividduang** [reply]
''oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000002-0000-0ff1-ce00-000000000000'.

**Comment by dividduang** [comment, marked as answer]
```python
def Email(to,cc_mail,subject, body, attachmentspath=None):

    creds = Credentials(
        username='xxxxx@xxxxxxcom',  # 账号
        password='xxxxxxxxxxx'  # 密码
    )
    account = Account(
        primary_smtp_address='*******@*********.com',  
        credentials=creds,
        autodiscover=True,
        access_type=DELEGATE
    )
    m = Message(
        account=account,
        cc_recipients=cc_mail,
        subject=subject,
        body=HTMLBody(body),
        to_recipients=[Mailbox(email_address=to)]
    )

im successful！！！！！！！！！！！
thx!!!love u
